### ISSUE TITLE
[improvement]: Petition to remove custom contexts from the standard from-pstn dialplan

### ISSUE THREAD

**Issue by lgaetz** [FreePBX/issue-tracker#524]
### FreePBX Version

FreePBX 17

### Improvement Description

I have long advocated in the fpbx forum against using the from-pstn-context. Due to the way that Asterisk combines contexts, each line put in from-pstn-custom overrides a generated dialplan line in from-pstn. This can cause issues that may not be immediately obvious when someone is testing their custom dialplan addition. 

Deep dive

Right now, the from-pstn context looks like this in 17
```
FreePBX17*CLI> dialplan show from-pstn
[ Context 'from-pstn' created by 'pbx_config' ]
  Include =>        'from-pstn-custom'                            [pbx_config]
  Include =>        'ext-did'                                     [pbx_config]
  Include =>        'ext-did-post-custom'                         [pbx_config]
  Include =>        'from-did-direct'                             [pbx_config]
  Include =>        'ext-did-catchall'                            [pbx_config]
```
from-pstn-custom is reserved for user defined dialplan - **problem context**
ext-did is used for incoming faxes and for inbound routes
ext-did-post-custom is reserved for user defined dialplan - **problem context**
from-did-direct is for incoming calls where the DID matches the extension number
ext-did-catchall place for inbound calls to land if there is no rule for handling them
```
FreePBX17*CLI> dialplan show ext-did
[ Context 'ext-did' created by 'pbx_config' ]
  'foo' =>          1. Noop(bar)                                  [extensions_additional.conf:3563]
  Include =>        'ext-did-custom'                              [pbx_config]
  Include =>        'ext-did-0001'                                [pbx_config]
  Include =>        'ext-did-0002'                                [pbx_config]
```
The foo extension is a surprise, what is this doing there? It doesn't hurt anything, but I can't think of any legit reason for this
ext-did-custom is reserved for user defined dialplan - **problem context**
ext-did-0001 - inbound faxes
ext-did 0002 - inbound routes

Recommendations

- Since 17 is now GA, I propose no major changes at this point for 17
- Announce that the following contexts are now deprecated, and will be removed in a future version, from-pstn-custom, ext-did-post-custom and ext-did-custom. Suggest they be removed sooner rather than later, target version 18
- investigate where the foo noop bar line came from in ext-did and remove if possible
- My first thought was to add gosub dialplan hooks in multiple places to replace each of the deprecated contexts, but that is actually quite difficult. Even if just adding for the most common use case, to hook an inbound route, it would have to be added in numerous places in ext-did-0002, so I think the only practical way of providing this utility is to just remove all the deprecated contexts, and if admins need to run custom dialplan on pstn calls, they can do so by creating their own inbound trunk context and ending that context with goto(from-pstn,${EXTEN},1)

**Comment by chrsmj** [comment]
Agreed that it can be confusing, but the flow is deterministic, and the CLI command `dialplan show exten@context` can help show you what is going to happen. Maybe parse that output and flag it for the user in the GUI with a message like "Hey, WYSIWYG break down ahead on this DID" ?  Improve documentation on the issue, generally ?

Because there are times when it is really nice to chmod 0400 /etc/asterisk/extensions_custom.conf and then walk away knowing that you helped shield the users from shots to the foot on the main number.

**Comment by lgaetz** [comment]
HI @chrsmj 

Not sure if your comment was for me or for others, I'm not confused. The three noted contexts (from-pstn-custom, ext-did-post-custom and ext-did-custom) when used by administrators in the way they are intended to be used will break things. In some cases the breakage is immediate and severe. In other cases, the breakage will not be evident in the initial testing phase. 

Instead of telling the user base not to use them, I submit that it's preferable that they don't exist.
